iwyu_output.cc

```diff
@@ -1776,27 +1776,53 @@ OutputLine PrintableIncludeOrForwardDeclareLine(
 
 typedef pair<int, string> LineSortKey;
 
+enum class LineSortOrdinal : int {
+  kPrecompiledHeader,
+  kAssociatedHeader,
+  kAssociatedInlineDefinitions,
+  kProjectLocalHeader,
+  kCHeader,
+  kCppHeader,
+  kOtherHeader,
+  kForwardDeclaration
+};
+
+template <typename T>
+typename std::underlying_type<T>::type underlying_cast(T val) {
+  static_assert(std::is_enum<T>::value, "underlying_cast only works with enums");
+  return static_cast<typename std::underlying_type<T>::type>(val);
+}
+
+LineSortOrdinal GetLineSortOrdinal(const OneIncludeOrForwardDeclareLine& line,
+                                   const set<string>& associated_quoted_includes,
+                                   const IwyuFileInfo* file_info) {
+  if (!line.IsIncludeLine())
+    return LineSortOrdinal::kForwardDeclaration;
+  if (file_info && file_info->is_pch_in_code())
+    return LineSortOrdinal::kPrecompiledHeader;
+
+  const auto& quotedInclude = line.quoted_include();
+  if (ContainsKey(associated_quoted_includes, quotedInclude)) {
+    if (EndsWith(quotedInclude, "-inl.h\""))
+      return LineSortOrdinal::kAssociatedInlineDefinitions;
+    return LineSortOrdinal::kAssociatedHeader;
+  }
+  if (EndsWith(quotedInclude, "\""))
+    return LineSortOrdinal::kProjectLocalHeader;
+  if (EndsWith(quotedInclude, ".h>"))
+    return LineSortOrdinal::kCHeader;
+  if (EndsWith(quotedInclude, ">"))
+    return LineSortOrdinal::kCppHeader;
+  return LineSortOrdinal::kOtherHeader;
```

Comment:
Hrmm. Good point. I'll take a look.

As to reason:
It's mostly to enable use in a system where you want to list higher-level dependencies first in the include ordering, which means that local project (i.e. usually quoted) headers would be stated first before external.

I'm happy to put it behind an option, which should also fix the Jenkins build as it looks like there are sample outputs I needed to update. Apologies, I just submitted my first PRs for IWYU today so I'm not as familiar with the validation tests.

Reply:
The reason you would want to list high level ones first is that it sometimes helps discover when you have unspecified dependencies between files or dependency inversions. If you include a high-level thing first that just always implicitly assumed a lower-level dependency was already included, you'll often get a failed build, which indicates that thing should have explicitly #included its dependencies itself so it could be used in greater isolation.